### DISCUSSION TITLE
Alternative method for joining clips

### DISCUSSION THREAD

**Discussion by petersampsonaudacity** [audacity/audacity#1497] [Ideas & Feature Requests]
I would like to propose an alternative mechanism be found and implemented for joining two adjacent clips.

Right now if two clips are immediately adjacent and the user left clicks on the shared boundary then the two clips are joined.  There is no indication, hover-text tool-tip, or visual cue - no affordance - to to the user that this joining will take place.  

We do know from the Forum that some users are surprised and unsettled by this.  Of course you can revert this with the use of Undo.

The IRL use case is trying to label the shared boundary pf two adjacent clips while retaining them as separate clips:
1) get some audio with two clips
2) move one of the clips so that the twp clips are immediately adjacent
3) Observe: you still have two distinct clips
4) try to create a point label at the shared boundary by left-clicking on the shared boundary
5) Observe: the two clips become joined into a single clip
6) Observe that it is almost impossible when clicking in step 4 to not move the mouse by a small amount (thus losing accuracy of clip sizes).

I would argue that at step 4 the left click should be reserve to only cursor positioning and that an alternative method be found to join clips:
a) perhaps just a simple right-click, or
b) a right, click context menu with a **Join Clips** command

See also Discussion #1499 
>Thought-piece: labels for merged clips #1499

@Tantacrul and @jounih  - as the UX/GUI experts on the team, what are your thoughts on this.

_The workaround BTW for this is to label what will become the shared boundary of the clip not to be moved prior to moving the other clip - but that assumes you know about the left-click clip joining._

**Comment by SteveDaulton** [comment]
> There is no indication, hover-text tool-tip, or visual cue - no affordance - to to the user that this joining will take place.

A tool tip could be helpful.




>     4\. try to create a point label at the shared boundary by left-clicking on the shared boundary

Not a problem, in fact extremely easy if a label track already exists. Once there is a label track, mouse pointer over the label  track and there's a yellow "snap" line to help  you position the label exactly at the clip boundary.
Could this issue be "solved" by providing a hint in the manual?

**Reply by petersampsonaudacity** [reply]
> A tool tip could be helpful.

Yes, indeed it would.

**Comment by Tantacrul** [comment]
100% agree we should use right-click for this. It will become more important once we build 'Smart clips' (the ability to widen or shorten a clip by dragging the edge of it)

**Reply by petersampsonaudacity** [reply]
> 100% agree we should use right-click for this. It will become more important once we build 'Smart clips' 
>(the ability to widen or shorten a clip by dragging the edge of it)

Thanks - and the suggestion of a tool-tip (for "discoverability") when hovering over a shared clip boundary still holds good if this change is made.

_The tooltip would be a great help not just not for new users - but also for existing clip users who are used to the left-click to join adjacent clips._

**Comment by werame** [comment]
While quick to execute, the left-click mergers are not exactly intuitive. They sorta violate the assumption that left click is mostly a selection action and non-destructive. Even more so that the one-click merging feature is active in pure select (F1) mode. I also dislike how easy is to accidentally move envelope control points by just left clicking in the wrong (F2) mode.

In some more expensive software (Audition) they do have a "razor" mode that allows the user to split clips by just left-clicking (there's no equivalent facility for one-click merging IIRC), but you do get **the mouse pointer changed** accordingly to a razor-looking thing in that mode, so you're forewarned that clicking is no longer a selecting operation.

So, perhaps changing the mouse pointer to something that suggests merging will take place should be done in Audacity, e.g. when the user hovers over the clip boundary. Likewise for one-click envelope moving, it would be better if the mouse pointer changed to indicate F2 mode. I don't know if wxWidgets supports this mouser-pointer changing cross platform (or at all) though.

Also, since there's now a "free mode" in Audacity with the [demise of F5 as the time tool](https://forum.audacityteam.org/viewtopic.php?f=46&t=121846), perhaps having a "razor mode" in Audacity for doing one-click splits would also be desirable and not hard to do. Although selecting first and pressing Ctrl+I isn't that big of a job.

**Comment by petersampsonaudacity** [comment]
There is already a GitHub issue relation to this - see:
> Remove the "Click to merge clips" line #2330

**Comment by werame** [comment]
Ok, so the Muse corp vision is to actually get rid of the one-click mergers altogether, rather than turn it into a mode.

**Reply by LWinterberg** [reply]
To give a bit more context as to why

Audacity has a long standing problem called "stuck in a mode" which has been worked on since 2017: https://wiki.audacityteam.org/wiki/ToDo:_Unstick_%27Stuck_In_a_Mode%27 

We generally are looking to continue in that direction, replacing special modes with other means of interacting with the project. For example, the envelope tool rather than being a mode which prevents you from selecting audio could instead be a "show envelopes on/off" toggle.

**Reply by petersampsonaudacity** [reply]
> For example, the envelope tool rather than being a mode which prevents you from selecting audio could instead be a "show envelopes on/off" toggle.

That would be great - as it would be another tool we could remove from the Tools toolbar (and then it wouldn't look like iy has an empty slot as now).   

_But more importantly it would stop folk accidental;y dropping into the bear-trap of accidentally clicking that tool button and not nowing how to revert to selection mode._

So @LWinterberg Leo, is this something we might see for 3.2.0?

**Reply by LWinterberg** [reply]
I believe the plan is to keep 3.2 focused on stackable effects and all the things attached to it (vst3, plugin manager, rewriting effects, etc), so probably not. But the entire tools toolbar is definitely the big thing our new designer is looking at right now :)
